### PR TITLE
Add Salt states exporter

### PR DESCRIPTION
**Pull request by guangyee**
Add a Salt states exporter. For example, to export the system description
to Salt states.

./bin/machinery export-salt --salt-dir /tmp/salt-foo foo --force

User is expected to copy the files in salt-dir to the Salt Master's
states dir. For example,

cp -r /tmp/salt-foo/foo-salt /srv/salt/foo

To apply the states.

salt 'foo' state.apply bar

### CODE REVIEW

File: lib/cli.rb [hunk 1]
```diff
@@ -616,6 +616,45 @@ def self.check_port_validity(port)
...
 
+    desc "Export system description as Salt states"
+    long_desc <<-LONGDESC
+      Export system description as Salt states.
+
+      The Salt states will be placed in a subdirectory at the given location
+      by the 'salt-dir' option.
+    LONGDESC
+    arg "NAME"
+    command "export-salt" do |c|
```
[the remark this answers is not in the document]
Comment (guangyee) [reply lib/cli.rb:627]: WTF hound?!

File: lib/salt_states.rb [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        repos_sls += "    - enabled: #{repo.enabled}\n"
+        repos_sls += "    - humanname: #{repo.alias}\n"
+        repos_sls += "    - refresh: #{repo.autorefresh}\n"
+        repos_sls += "    - gpgcheck: #{repo.gpgcheck}\n\n"
+      end
+    end
+    repos_sls
+  end
+
+  def service_states
```
[the remark this answers is not in the document]
Comment (guangyee) [reply lib/salt_states.rb:378]: Unavoidable

File: lib/salt_states.rb [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+          if member == user
+            groups.append(group.name)
+            break
+          end
+        end
+      end
+      groups
+    end
+
+    def add_user(user, groups)
```
[the remark this answers is not in the document]
Comment (guangyee) [reply lib/salt_states.rb:199]: Bullshit

File: lib/salt_states.rb [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+          archive.extracted:
+            - name: /
+            - source: salt://#{export_name}/#{filepath}
+        #{options}
+      END_SALT_STATE
+    end
+
+    # NOTE(gyee): tarballs_dir is relative to the /srv/salt/, where the Salt
+    # state files are located.
+    def unmanaged_file_states(output_location)
```
[the remark this answers is not in the document]
Comment (guangyee) [reply lib/salt_states.rb:271]: Bullshit

File: lib/salt_states.rb [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+        @system_description.unmanaged_files.export_files_as_tarballs(export_dir)
+        Dir["#{export_dir}/**/*.tgz"].sort.each do |path|
+          filepath = salt_tarballs_dir + path.gsub(export_dir, "")
+          unmanaged_files_sls += add_unmanaged_file(filepath, filter_options)
+        end
+      end
+      unmanaged_files_sls
+    end
+
+    def managed_file_scope_filter(scope)
```
[the remark this answers is not in the document]
Comment (guangyee) [reply lib/salt_states.rb:289]: Bullshit

File: lib/salt_states.rb [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        end
+      end
+    end
+    services_sls
+  end
+
+  # TODO(gyee):
+  #
+  # - network?
+end
```
[the remark this answers is not in the document]
Comment (guangyee) [reply lib/salt_states.rb:504]: Huh?

**Review by jeremy-moffitt — APPROVED** [review]
There's a lot of fixme/todos here, but this looks good overall. I recommend at least one other person review before merging if possible as Ruby is not my strongsuit.

**Review by chajain — APPROVED** [review]
Looks good to me

**Pull request merged**